src/analysis/complex/upper_half_plane.lean

```diff
@@ -0,0 +1,172 @@
+/-
+Copyright (c) 2021 Alex Kontorovich and Heather Macbeth and Marc Masdeu. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Alex Kontorovich, Heather Macbeth, Marc Masdeu
+-/
+
+import linear_algebra.special_linear_group
+import analysis.complex.basic
+import group_theory.group_action.defs
+
+/-!
+# The upper half plane and its automorphisms
+
+This file defines `upper_half_plane` to be the upper half plane in `ℂ`.
+
+We furthermore equip it with the structure of an `SL(2,ℝ)` action by
+fractional linear transformations.
+-/
+
+noncomputable theory
+
+open matrix matrix.special_linear_group
+
+open_locale classical big_operators
+
+local attribute [instance] fintype.card_fin_even
+
+@[ext] structure upper_half_plane :=
```

Comment:
Is there an advantage to this definition vs just declaring it as a subtype?

Reply:
We use projection notation a lot in subsequent work (see the branch modular3) -- `z.im`, `z.im_pos`, etc.  Does the subtype definition permit that?  I think that's of much more importance in user-friendliness than the add_submonoid structure.

Reply:
Yes, I think that should work just fine

Reply:
Also it obviously does not have an add_comm_monoid structure, I missed the fact this does not include the real line!

Reply:
We made the change, thanks for the suggestion!